### PR TITLE
Implement builtin zip

### PR DESCRIPTION
**Pull request by S-YOU**
Let me know if my approach is not right.
More test case has to be done. Will do it my time tomorrow.

### CODE REVIEW

File: runtime/builtin_types.go [hunk 1]
```diff
@@ -552,6 +553,37 @@ func builtinUniChr(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
 	return NewUnicodeFromRunes([]rune{rune(i)}).ToObject(), nil
 }
 
+func builtinZip(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+	seqs := [][]*Object{}
+	lenHint := math.MaxInt64
+	for _, arg := range args {
+		if raised := checkFunctionArgs(f, "zip", Args{arg}, ObjectType); raised != nil {
```
Comment (trotterdylan) [inline runtime/builtin_types.go]: This check will never fail so just remove it.

File: runtime/builtin_types.go [hunk 2]
```diff
@@ -552,6 +553,37 @@ func builtinUniChr(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
...
+			return nil, raised
+		}
+		seq, raised := seqNew(f, Args{arg})
```
Comment (trotterdylan) [inline runtime/builtin_types.go]: It's expensive to allocate all these slices, so I would recommend creating an iterator for each arg and calling next over all the iterators until one is exhausted.

That means you won't get a lenHint which makes it a little more expensive to build the result list, but you can grow the list slice by doubling its capacity every time to minimize the number of reallocations.

File: runtime/builtin_types.go [hunk 3]
```diff
@@ -552,6 +553,37 @@ func builtinUniChr(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
...
+			lenHint = len(seq)
+		}
+	}
+	list := make([]*Object, lenHint)
+	for i := 0; i < lenHint; i++ {
+		tuple := make([]*Object, len(args))
+		for j := 0; j < len(args); j++ {
+			tuple[j] = seqs[j][i]
+		}
+		t := toTupleUnsafe(newObject(TupleType))
```
Comment (trotterdylan) [inline runtime/builtin_types.go]: `t := NewTuple(tuple...)`

File: runtime/builtin_types.go [hunk 4]
```diff
@@ -552,6 +553,37 @@ func builtinUniChr(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
...
+		t.elems = tuple
+		list[i] = t.ToObject()
+	}
+	l := toListUnsafe(newObject(ListType))
```
Comment (trotterdylan) [inline runtime/builtin_types.go]: `l := NewList(list...)`

**Comment by S-YOU** [comment]
@trotterdylan, rewrote most part to reflect iter. PTAL.
I didnt manually make slice grow, but just using append.

**Comment by S-YOU** [comment]
@trotterdylan, updated code to use slice growth by double, by using 1 dimensional slice.
Please take a look.

**Comment by S-YOU** [comment]
Thanks for the implementation @trotterdylan. It just works :+1:

File: runtime/builtin_types.go [hunk 5]
```diff
@@ -565,6 +565,40 @@ func builtinUniChr(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
...
+			elem, raised := Next(f, iter)
+			if raised != nil {
+				if raised.isInstance(StopIterationType) {
+					break Outer
+				}
+				return nil, raised
+			}
+			elems[i] = elem
+		}
+		// NOTE: append does the capacity doubling logic as needed
```
Comment (trotterdylan) [inline runtime/builtin_types.go]: Remove this comment.

File: runtime/builtin_types.go [hunk 6]
```diff
@@ -565,6 +565,40 @@ func builtinUniChr(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
...
+	}
+
+Outer:
+	for {
+		elems := make([]*Object, argc)
+		for i, iter := range iters {
+			elem, raised := Next(f, iter)
+			if raised != nil {
+				if raised.isInstance(StopIterationType) {
+					break Outer
```
Comment (trotterdylan) [inline runtime/builtin_types.go]: You'll also need to clear the exception state on the frame by calling `f.RestoreExc(nil, nil)` similar to [here](https://github.com/google/grumpy/blob/master/runtime/core.go#L122).

Comment (S-YOU) [inline runtime/builtin_types.go]: Thank you for explanation, I noticed that line somewhere, but didn't understand.

File: runtime/builtin_types.go (hunk 5 above)
Comment (S-YOU) [inline runtime/builtin_types.go]: Done

**Comment by trotterdylan** [comment]
Thanks for implementing this! Merging now.

**Pull request merged**